fetcher/fetcher.go

```diff
@@ -0,0 +1,249 @@
+func New(numTxs int, concurrency int, im state.Immutable) *Fetcher {
+	f := &Fetcher{
+		im:          im,
+		cache:       make(map[string]*fetchData, numTxs),
+		keysToFetch: make(map[string]*keyData),
+		txnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task, numTxs),
+		stop:        make(chan struct{}),
+		numTxs:      numTxs,
+	}
+	for i := 0; i < concurrency; i++ {
+		f.wg.Add(1)
+		go f.runWorker()
+	}
+	return f
+}
+
+// Workers fetch individual keys
+func (f *Fetcher) runWorker() {
+	defer f.wg.Done()
+
+	for {
+		select {
+		case t, ok := <-f.fetchable:
+			if !ok {
+				return
+			}
+
+			// Allows concurrent reads to cache.
+			if ok = f.check(t); ok {
+				continue
+			}
+
+			// Fetch from disk once if key isn't already in cache
+			v, err := f.im.GetValue(t.ctx, []byte(t.key))
+			if errors.Is(err, database.ErrNotFound) {
+				f.update(t.key, nil, false, 0)
+				continue
+			} else if err != nil {
+				f.stopOnce.Do(func() {
+					f.err = err
+					close(f.stop)
+				})
+				return
+			}
+
+			numChunks, ok := keys.NumChunks(v)
+			if !ok {
+				f.stopOnce.Do(func() {
+					f.err = ErrInvalidKeyValue
+					close(f.stop)
+				})
+				return
+			}
+			f.update(t.key, v, true, numChunks)
+		case <-f.stop:
+			return
+		}
+	}
+}
+
+// Checks if a key is in the cache. If it's not in cache, check
+// if we were the first to request the key.
+func (f *Fetcher) check(t *task) bool {
+	f.keyLock.RLock()
+	defer f.keyLock.RUnlock()
+
+	_, exists := f.cache[t.key]
+	if !exists {
+		queue := f.keysToFetch[t.key].queue
+		if len(queue) > 0 && queue[0] != t.id {
+			// Fetch other keys instead
+			return true
+		}
+	}
+	return exists
+}
+
+// Updates the cache and dependencies
+func (f *Fetcher) update(k string, v []byte, exists bool, chunks uint16) {
+	f.keyLock.Lock()
+	defer f.keyLock.Unlock()
+
+	// Puts a key that was fetched from disk into cache
+	f.cache[k] = &fetchData{v, exists, chunks}
+
+	f.txnLock.Lock()
+	for _, id := range f.keysToFetch[k].queue {
+		f.txnsToFetch[id].Done() // Notify all other txs
+	}
+	f.txnLock.Unlock()
+	f.keysToFetch[k].queue = nil
+	f.keysToFetch[k].fetched = true
+}
+
+// Lookup enqueues keys for the workers to fetch
+func (f *Fetcher) Lookup(ctx context.Context, txID ids.ID, stateKeys state.Keys) *sync.WaitGroup {
+	wg := &sync.WaitGroup{}
+	f.txnLock.Lock()
+	f.txnsToFetch[txID] = wg
+	f.txnLock.Unlock()
+
+	f.keyLock.Lock()
+	tasks := make([]*task, 0, len(stateKeys))
+	for k := range stateKeys {
+		if _, ok := f.keysToFetch[k]; !ok {
+			f.keysToFetch[k] = &keyData{fetched: false, queue: make([]ids.ID, 0)}
+		}
+		// Don't requeue keys that were already fetched
+		if f.keysToFetch[k].fetched {
+			continue
+		}
+		f.keysToFetch[k].queue = append(f.keysToFetch[k].queue, txID)
+		t := &task{
+			ctx: ctx,
+			id:  txID,
+			key: k,
+		}
+		tasks = append(tasks, t)
+	}
+	f.keyLock.Unlock()
+
+	if done := f.stopped(); done {
+		return wg // counter is zero
```

Comment:
Should we change the return value of the function to also include an error?

Reply:
I think it might just be easier to treat this as an invariant (which we satisfy in our usage) that says, "Don't call Lookup after calling Stop".

Then you can remove this extra locking.